cpp/src/coll/allreduce.cpp

```diff
@@ -0,0 +1,122 @@
+    std::shared_ptr<Communicator> comm,
+    std::shared_ptr<ProgressThread> progress_thread,
+    OpID op_id,
+    BufferResource* br,
+    std::shared_ptr<Statistics> statistics,
+    ReduceKernel reduce_kernel,
+    std::function<void(void)> finished_callback
+)
+    : comm_{std::move(comm)},
+      progress_thread_{std::move(progress_thread)},
+      br_{br},
+      statistics_{std::move(statistics)},
+      reduce_kernel_{std::move(reduce_kernel)},
+      finished_callback_{std::move(finished_callback)},
+      gatherer_{comm_, progress_thread_, op_id, br_, statistics_} {
+    RAPIDSMPF_EXPECTS(
+        static_cast<bool>(reduce_kernel_),
+        "AllReduce requires a valid ReduceKernel at construction time"
+    );
+}
+
+AllReduce::~AllReduce() = default;
+
+void AllReduce::insert(std::uint64_t sequence_number, PackedData&& packed_data) {
+    nlocal_insertions_.fetch_add(1, std::memory_order_relaxed);
+    gatherer_.insert(sequence_number, std::move(packed_data));
+}
+
+void AllReduce::insert_finished() {
+    gatherer_.insert_finished();
+}
+
+bool AllReduce::finished() const noexcept {
+    return gatherer_.finished();
+}
+
+std::vector<PackedData> AllReduce::wait_and_extract(std::chrono::milliseconds timeout) {
+    // Block until the underlying allgather completes, then perform the reduction locally
+    // (exactly once).
+    if (!reduced_computed_.load(std::memory_order_acquire)) {
+        auto gathered =
+            gatherer_.wait_and_extract(AllGather::Ordered::YES, std::move(timeout));
+        reduced_results_ = reduce_all(std::move(gathered));
+        reduced_computed_.store(true, std::memory_order_release);
+        if (finished_callback_) {
+            finished_callback_();
+        }
+    }
+    return std::move(reduced_results_);
+}
+
+bool AllReduce::is_ready() const noexcept {
+    return reduced_computed_.load(std::memory_order_acquire) || gatherer_.finished();
+}
+
+std::vector<PackedData> AllReduce::reduce_all(std::vector<PackedData>&& gathered) {
+    auto const nranks = static_cast<std::size_t>(comm_->nranks());
+    auto const total = gathered.size();
+
+    if (total == 0) {
+        return {};
+    }
+
+    RAPIDSMPF_EXPECTS(
+        nranks > 0, "AllReduce requires a positive number of ranks", std::runtime_error
+    );
+    RAPIDSMPF_EXPECTS(
+        total % nranks == 0,
+        "AllReduce expects each rank to contribute the same number of messages",
+        std::runtime_error
+    );
+
+    auto const n_local =
+        static_cast<std::size_t>(nlocal_insertions_.load(std::memory_order_acquire));
+    auto const n_per_rank = total / nranks;
+
+    // We allow non-uniform insertion counts across ranks but require that the local
+    // insertion count matches the per-rank contribution implied by the gather.
+    RAPIDSMPF_EXPECTS(
+        n_local == 0 || n_local == n_per_rank,
+        "AllReduce local insertion count does not match gathered contributions per rank",
+        std::runtime_error
+    );
```

Comment:
This makes no sense. Either, everyone has to insert exactly `k` entries and the `i`th result is computed by reducing over the `PackedData` pieces that are sliced from `[i::k]`, or everyone inserts different numbers of entries (as is claimed is being done here) and then you can't do that.